generate_normalized_intensity_data_for_swabs_or_deliveries.py

```diff
@@ -0,0 +1,212 @@
+"""
+This script generates genotype call (.gtc) files
+from raw intensity data (.idats) and extracts
+NormR and NormTheta values at all array markers
+for a list of dogs (swab codes), or all dogs in specified deliveries.
+
+The operations the script performs require a lot of disk space,
+and it has not been tested on a lower RAM machine,
+so it is recommended that it be run on an ec2 instance instead of locally.
+
+TODO: add functionality to restrict data to a subset of markers rather than entire array
+
+Example usage:
+
+`python generate_normalized_intensity_data_for_swabs_or_deliveries.py
+    --delivery-names embark_2021-12-26_0609
+    --cluster-file-s3-path s3://illumina-embark-data/Cluster_Files_Updated_Monthly/Embark_2021_260k_20063270_A1_20211122.egt
+    --beadpool-manifest-s3-path s3://illumina-embark-data/beadpool-manifests/Embark_2021_260k_20063270_A1.bpm
+    --output-dir /home/ubuntu/idat_to_gtc_test
+`
+
+"""
+
+import argparse
+import functools
+import os
+import shutil
+from typing import List
+from multiprocessing import Pool, cpu_count
+from pathlib import Path
+import pandas as pd
+import numpy as np
+from IlluminaBeadArrayFiles import GenotypeCalls, BeadPoolManifest
+from pybark import db, s3, shell
+
+POOL_SIZE = cpu_count() * 2
+env_folder = s3.env_folder("prod")
+s3_bucket = "illumina-embark-data"
+
+
+def _get_delivery_names_for_swab_codes(swab_codes: List[str]) -> pd.DataFrame:
+    results_df = pd.DataFrame(
+        db.run_query(
+            ENVIRONMENT=db.get_environment(),
+            sql="SELECT illumina_delivery_name, swab_code FROM genotypes WHERE swab_code IN %(swab_codes)s",
+            substitutions={"swab_codes": tuple(swab_codes)},
+        ),
+    )
+    return results_df
+
+
+def _get_swab_codes_for_delivery_names(delivery_names: List[str]) -> pd.DataFrame:
+    results_df = pd.DataFrame(
+        db.run_query(
+            ENVIRONMENT=db.get_environment(),
+            sql="SELECT illumina_delivery_name, swab_code FROM genotypes WHERE illumina_delivery_name IN %(delivery_names)s",
+            substitutions={"delivery_names": tuple(delivery_names)},
+        ),
+    )
+    return results_df
+
+
+def _download_idats(output_dir: Path, record: dict) -> None:
+    swab_code = record["swab_code"]
+    sentrix_id = record["sentrix_id"]
+    sentrix_position = record["sentrix_position"]
+    delivery_name = record["illumina_delivery_name"]
+    for color in "Grn", "Red":
+        idat_s3_path = s3.construct_s3_url(
+            bucket=s3_bucket,
+            key=f"{env_folder}unzipped/{delivery_name}/{swab_code}/IDATs/{sentrix_id}/{sentrix_id}_{sentrix_position}_{color}.idat",
+        )
+        s3.download_file(
+            idat_s3_path, f"{output_dir}/idats/{sentrix_id}_{sentrix_position}_{color}.idat"
+        )
+
+
+def _make_gtcs_from_idats_for_dir(delivery_data_dir: Path, beadpool_manifest_path: Path, cluster_file_path: Path) -> None:
```

Comment:
Do we get GTC files from Illumina already? If so, would it make sense to use those instead of re-generate them here?

Reply:
We don't, we only receive .idats and .ped files.